editor/reducer.js

```diff
@@ -454,6 +422,33 @@ export function preferences( state = STORE_DEFAULTS.preferences, action ) {
 				...state,
 				mode: action.mode,
 			};
+		case 'INSERT_BLOCKS':
+			// record the block usage and put the block in the recently used blocks
+			let blockUsage = state.blockUsage;
+			let recentlyUsedBlocks = [ ...state.recentlyUsedBlocks ];
+			action.blocks.forEach( ( block ) => {
+				const uses = ( blockUsage[ block.name ] || 0 ) + 1;
+				blockUsage = omit( blockUsage, block.name );
+				blockUsage[ block.name ] = uses;
+				recentlyUsedBlocks = [ block.name, ...without( recentlyUsedBlocks, block.name ) ].slice( 0, maxRecent );
+			} );
+			return {
+				...state,
+				blockUsage,
+				recentlyUsedBlocks,
+			};
+		case 'SETUP_EDITOR':
+			// initially populate the recently used blocks with blocks the user most frequently uses
+			const commonBlocks = getBlockTypes()
+				.filter( ( blockType ) => 'common' === blockType.category )
+				.map( ( blockType ) => blockType.name );
+			const frequentlyUsedBlocks = keys( state.blockUsage ).sort( ( a, b ) => state.blockUsage[ b ] - state.blockUsage[ a ] );
+			return {
+				...state,
+				recentlyUsedBlocks: frequentlyUsedBlocks
+					.concat( difference( commonBlocks, frequentlyUsedBlocks ) )
+					.slice( 0, maxRecent ),
+			};
```

Comment:
Why do we need to store `blockUsage` in state and do this initialization of `recentlyUsedBlocks` since `recentlyUsedBlocks` is also persisted?

Reply:
I wanted some feedback on this... when we start a new post, we want to populate the recently used blocks based on what blocks the user uses most frequently, so their most common stuff is there ready for them, instead of perhaps the twitter embed they finished their last post with, if you see what I mean. So, the initial state of `recentlyUsedBlocks` relies on `blockUsage`, but I couldn't get to `blockUsage` from a separate reducer, so it moved into preferences. Open to suggestions here!

Reply:
I don't have a strong preference on the algorithm itself, but in this case, I'd assume that `INSERT_BLOCKS` works similarly. If the newly created block is not among the most used blocks, don't add it. I'm thinking we should be consistent.

Also, when do we start counting the usage, should we fix a threshold or keep these numbers grow infinitely?

Reply:
> If the newly created block is not among the most used blocks, don't add it.

Then how do we get any most used blocks? We'd never be able to add any, or they'd never be able to change. Or do you mean that if a block is already in the most used, don't increment?

Either way, we're just recording an int, so if I remember correctly, we can record up 
 to `Number.MAX_SAFE_INTEGER` which is 9007199254740991. I'm not sure most of our users will ever insert that many of a single block?

Reply:
I'm still not sure how that would work, because we want to record how frequently blocks are used so that we can generate a list of the most frequently used blocks. If we didn't increment the usage, then we wouldn't end up with useful data... or am I understanding this wrong? I'm not sure what you're getting at here.

Reply:
Perhaps I'm not explaining the problem I'm solving very well..

I want this to do three things:

1) Keep track of what blocks a user uses most frequently.
2) Use the usage data to populate the "recent" tab with their most frequently used blocks on editor setup
3) During an editing session, have the most recently used blocks appear in the "recent" tab.

As far as I can see, that requires we keep track of how many times a user uses blocks, and either have a list of recently used ones, or keep `last_used` data. Personally, I'd rather go for the latter, as two data structures seems wrong when it's really just two aspects of block usage, but as you want to have a memoized selector based on that data, that becomes difficult. Why do you want to memoize? Is it to avoid getting the concrete blocks each time?

Reply:
I didn't say don't increment the usage :). I just use one unique technique: 

 - increment usage on insert and use the selector above
 - or do not save the usage at all and rely on the last inserted blocks (like before)

Reply:
So what you're suggesting is that the user have a list of rencent blocks shown but when you refresh the page, this list changes completely because we'll rely on the usage to recompute this list. Why this difference? I just think we should go for consistency and keep the same list when refreshing.

Reply:
Yes, that's what this does, and why I wanted feedback on it :smile: 

Perhaps we do just want to persist the recently used list itself? If someone uses a block frequently, it's unlikely to drop out of the list, although it may change position, as we put the most recently used ones at the top. Or perhaps it's always a list of the most frequently used blocks, ordered by how often they're used?

Reply:
Yep, I don't know the answer. My suggestion was being consistent across reloads but let's what others think @karmatosed @jasmussen